### PR TITLE
Creating NuGet.Localization.nupkg for delivering localized components to dotnet cli

### PR DESCRIPTION
**Pull request by mishra14**
Adding a script `scripts/cibuild/CreateLocalizedPackage.ps1` that is invoked by VSTS cli to create a localization package containing all the satellite assemblies for our `nupkgs`.

The script invokes a target in loc.proj which then invokes targets per project to get all localized dlls. These dlls are then placed in a common location. 

I have also added a new `nuspec` file to pack the package.

I am going to work with someone from the cli team to verify this. But please take a look in the mean time.

Once this is merged I will add the VSTS step to the official build definition.

### CODE REVIEW

**Review by emgarten — CHANGES_REQUESTED** [review]
move to MSBuild

File: build/common.targets [hunk 1]
```diff
@@ -196,6 +196,36 @@
...
+
+      <Output
+          TaskParameter="TargetOutputs"
+          ItemName="LocalizedFiles" />
+    </MSBuild>
+  </Target>
+
+  <Target Name="GetLocalizedFilesFromProject" Returns="@(LocalizedFilesFromProject)" Condition=" '$(PackProject)' == 'true' ">
+    <ItemGroup>
+      <_AllResourceDlls Include="$(OutputPath)**\$(AssemblyName).resources.dll"/>
```
Comment (rohit21agrawal) [inline build/common.targets]: Isn't this item group already being used in a different target?

File: build/common.targets [hunk 2]
```diff
@@ -196,6 +196,36 @@
...
+      <LocalizedFilesFromProject Include="@(_AllResourceDlls)" Condition="'@(_AllResourceDlls)'!=''">
```
Comment (rohit21agrawal) [inline build/common.targets]: Don't think this condition makes sense for item groups

**Comment by rohit21agrawal** [comment]
If you can wait until Monday, i will show you a couple more tricks to clean this up

**Review by rohit21agrawal — CHANGES_REQUESTED** [review]

File: build/common.targets (hunk 2 above)
Comment (mishra14) [reply build/common.targets]: In case the resource files are not present, the this condition prevents failures.

File: build/common.targets [hunk 3]
```diff
@@ -196,6 +196,36 @@
...
+    GetLocalizedFiles - gathers the list of files that have been
+    localized from each project
+    ============================================================
+  -->
+  <!--Target is invoked by build\loc.proj\MoveLocalizedFilesToLocalizedArtifacts to get all the localized files from all projects-->
+  <Target Name="GetLocalizedFiles" DependsOnTargets="GetTargetFrameworkSet" Returns="@(LocalizedFiles)" Condition=" '$(Shipping)' == 'true' ">
+    <MSBuild
+      Projects="$(MSBuildProjectFullPath)"
+      Targets="GetLocalizedFilesFromProject"
+      Properties="TargetFramework=%(ProjectTargetFrameworkEntries.Identity);
```
Comment (rohit21agrawal) [inline build/common.targets]: Why do you need to do an inner build?

Comment (rohit21agrawal) [reply build/common.targets]: All the localised files are already present in $(LocalizationRootDirectory) , this move is redundant

**Review by emgarten — APPROVED** [review]

**Review by rohit21agrawal — CHANGES_REQUESTED** [review]

File: build/common.targets (hunk 3 above)
Comment (mishra14) [reply build/common.targets]: Where, in code, are those files moved? I need them in the form of `project_name/framework/culture/resource_dll`. This is what this target helps with.

File: build/common.targets [hunk 4]
```diff
@@ -196,6 +196,36 @@
...
+
+      <Output
+          TaskParameter="TargetOutputs"
+          ItemName="LocalizedFiles" />
+    </MSBuild>
+  </Target>
+
+  <Target Name="GetLocalizedFilesFromProject" Returns="@(LocalizedFilesFromProject)" Condition=" '$(PackProject)' == 'true' ">
+    <ItemGroup>
+      <LocalizedFiles Include="$(OutputPath)**\$(AssemblyName).resources.dll"/>
```
Comment (rohit21agrawal) [inline build/common.targets]: start with _ since these are internal itemgroups

File: build/build.proj [hunk 5]
```diff
@@ -167,6 +167,46 @@
...
+  <Target Name="MoveLocalizedFilesToLocalizedArtifacts">
+    <MSBuild
+      Projects="@(SolutionProjectsWithoutVSIX)"
+      Properties="BuildProjectReferences=false;
+                  Configuration=$(Configuration)"
+      Targets="GetLocalizedFiles">
+
+      <Output
+          TaskParameter="TargetOutputs"
+          ItemName="LocalizedFiles" />
```
Comment (rohit21agrawal) [inline build/build.proj]: LocalizedFilesForNupkg

File: build/build.proj [hunk 6]
```diff
@@ -167,6 +167,46 @@
...
+    </MSBuild>
+
+    <ItemGroup>
+      <LocalizedFilePaths Include="@(LocalizedFiles)">
+        <TargetPath>$(ArtifactsDirectory)LocalizedFiles\%(LocalizedFiles.RelativeTargetPath)</TargetPath>
```
Comment (rohit21agrawal) [inline build/build.proj]: change this to msbuild variable

File: build/build.proj [hunk 7]
```diff
@@ -167,6 +167,46 @@
...
+    <Copy SourceFiles="@(LocalizedFilePaths->'%(Identity)')" DestinationFiles="@(LocalizedFilePaths->'%(TargetPath)')"/>
+    <Message Text="Localized Files Moved: %(LocalizedFilePaths.TargetPath)" Importance="High"/>
+  </Target>
+
+  <!--
+    ============================================================
+    Generate NuGet.Localized.nupkg.
+    ============================================================
+  -->
+  <Target Name="GenerateLocalizedPackage" DependsOnTargets="MoveLocalizedFilesToLocalizedArtifacts" Condition=" '$(IsXPlat)' != 'true' ">
```
Comment (rohit21agrawal) [inline build/build.proj]: change to $(PackProject) == true

**Review by rohit21agrawal — CHANGES_REQUESTED** [review]

File: build/Nuspec/NuGet.Localization.nuspec [hunk 8]
```diff
@@ -0,0 +1,19 @@
+<?xml version="1.0" encoding="utf-8"?>
```
Comment (rohit21agrawal) [inline build/Nuspec/NuGet.Localization.nuspec]: put this file under src as well. Create a folder for the project name and put it under there.

File: build/common.targets [hunk 9]
```diff
@@ -196,6 +196,35 @@
 
   <!--
     ============================================================
+    GetLocalizedFiles - gathers the list of files that have been
+    localized from each project
+    ============================================================
+  -->
+  <!--Target is invoked by build\loc.proj\MoveLocalizedFilesToLocalizedArtifacts to get all the localized files from all projects-->
+  <Target Name="GetLocalizedFilesForLocalizationPackage" Returns="@(_LocalizedFilesForLocalizationPackage)" Condition=" '$(Shipping)' == 'true' ">
```
Comment (rohit21agrawal) [inline build/common.targets]: this should also be PackProject == true

**Comment by emgarten** [comment]
:shipit:

**Comment by mishra14** [comment]
Had an offline discussion with @emgarten. merging this in to get this into CLI ASAP.

**Pull request merged**